src/Stratis.Bitcoin.Features.PoA.IntegrationTests/Tools/TestPoAMiner.cs

```diff
@@ -0,0 +1,81 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using NBitcoin;
+using Stratis.Bitcoin.Connection;
+using Stratis.Bitcoin.Consensus;
+using Stratis.Bitcoin.Consensus.Validators;
+using Stratis.Bitcoin.Tests.Common;
+using Stratis.Bitcoin.Features.Miner;
+using Stratis.Bitcoin.Features.Wallet.Interfaces;
+using Stratis.Bitcoin.Interfaces;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Bitcoin.Features.PoA.IntegrationTests.Tools
+{
+    public class TestPoAMiner : PoAMiner
+    {
+        public bool FastMiningEnabled { get; private set; } = false;
+
+        private readonly DateTimeProvider timeProvider;
+
+        private CancellationTokenSource cancellationSource;
+
+        public TestPoAMiner(
+            IConsensusManager consensusManager,
+            IDateTimeProvider dateTimeProvider,
+            Network network,
+            INodeLifetime nodeLifetime,
+            ILoggerFactory loggerFactory,
+            IInitialBlockDownloadState ibdState,
+            BlockDefinition blockDefinition,
+            SlotsManager slotsManager,
+            IConnectionManager connectionManager,
+            PoABlockHeaderValidator poaHeaderValidator,
+            FederationManager federationManager,
+            IIntegrityValidator integrityValidator,
+            IWalletManager walletManager) : base(consensusManager, dateTimeProvider, network, nodeLifetime, loggerFactory, ibdState, blockDefinition, slotsManager,
+                connectionManager, poaHeaderValidator, federationManager, integrityValidator, walletManager)
+        {
+            this.timeProvider = dateTimeProvider as DateTimeProvider;
+            this.cancellationSource = new CancellationTokenSource();
+        }
+
+        public void EnableFastMining()
+        {
+            this.FastMiningEnabled = true;
+            this.cancellationSource.Cancel();
+        }
+
+        public void DisableFastMining()
+        {
+            this.FastMiningEnabled = false;
+            this.cancellationSource = new CancellationTokenSource();
+        }
+
+        protected override async Task WaitBeforeCanMineAsync(int delayMs, CancellationToken cancellation = default(CancellationToken))
+        {
+            if (this.FastMiningEnabled)
+            {
+                TimeSpan span = (TimeSpan)this.timeProvider.GetMemberValue("adjustedTimeOffset");
```

Comment:
This is really ugly, instead I would suggest to cast to a new DTP type with those methods publicly available.
but its test to no  big deal

Reply:
I'm not sure if we should change production code in favor of tests. 
Also we've agreed long time ago to use reflection over modifying production code.

Reply:
No the new DTP can be a test class that you inject instead of the normal provider

Reply:
Ok, fixed. 
But I had to change adjustedTimeOffset to be protected, which is not good, but ok.